Report `packages` at top level with file level `package_manifests`

AyanSinhaMahapatra opened the pull request:
Fixes #2694 

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

AyanSinhaMahapatra committed:
Modify file level attribute packages to package_manifests

See #2694

Signed-off-by: Ayan Sinha <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify tests and test expectations

Modifies tests and test expectations for the change of
file level attribute `packages` to `package_manifests`.

Signed-off-by: Ayan Sinha <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into 2098-top-level-packages

Signed-off-by: Ayan Sinha <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@pombredanne All green! Ready to review.

AyanSinhaMahapatra committed:
Add top level `packages` codebase attribute

Adds `packages` codebase attribute and temporarily populate with
all package manifests detected in the codebase.

Signed-off-by: Ayan Sinha <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify test expectations for top level `packages`

Signed-off-by: Ayan Sinha <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into 2098-top-level-packages

Signed-off-by: Ayan Sinha <ayansmahapatra@gmail.com>

pombredanne commented on src/packagedcode/plugin_package.py:
```diff
@@ -92,6 +94,21 @@ def process_codebase(self, codebase, **kwargs):
+        summarize_codebase_package_manifests(codebase, **kwargs)
+
+
+def summarize_codebase_package_manifests(codebase, **kwargs):
```
What about rather naming this `create_packages` or possibly `create_packages_from_manifests`

pombredanne commented on src/summarycode/plugin_consolidate.py:
```diff
@@ -98,21 +98,21 @@ def to_dict(self, **kwargs):
 
 
 @attr.s
-class ConsolidatedPackage(object):
```
Keep this whole consolidation code as-is. We may merge or remove some of it at a later stage, especially since getting package files would be doing double duty with some this.

pombredanne commented on tests/formattedcode/data/csv/livescan/expected.csv:
```diff
@@ -1,16 +1,16 @@
 Resource,type,name,base_name,extension,size,date,sha1,md5,sha256,mime_type,file_type,programming_language,is_binary,is_text,is_archive,is_media,is_source,is_script,percentage_of_license_text,files_count,dirs_count,size_count,scan_errors,license_expression,license__key,license__score,license__name,license__short_name,license__category,license__is_exception,license__is_unknown,license__owner,licens ...
-json2csv.rb,file,json2csv.rb,json2csv,.rb,912,2020-04-27,1236469a06a2bacbdd8e172ad718482af5b0a936,1307c281e0b153202e291b217eab85d5,12ba215313981dbe810d9ed696b7cc753d97adfcc26eba1e13f941dc7506aa4e,text/x-script.python,"Python script, ASCII text executable",Ruby,False,True,False,False,True,True,62.04,0,0,0,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
+json2csv.rb,file,json2csv.rb,json2csv,.rb,912,2021-09-09,1236469a06a2bacbdd8e172ad718482af5b0a936,1307c281e0b153202e291b217eab85d5,12ba215313981dbe810d9ed696b7cc753d97adfcc26eba1e13f941dc7506aa4e,text/x-script.python,"Python script, ASCII text executable",Ruby,False,True,False,False,True,True,62.04,0,0,0,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
```
For just a date change, it is best to avoid committing an updated fixture

pombredanne commented on tests/licensedcode/data/plugin_license/package/package.expected.json:
```diff
@@ -17,6 +18,7 @@
       }
     }
   ],
+  "packages": [],
```
IMHO we are missing the packages

pombredanne commented on tests/packagedcode/data/plugin/about-package-expected.json:
```diff
@@ -134,7 +229,7 @@
     {
       "path": "appdirs.ABOUT-expected",
       "type": "file",
-      "packages": [],
```
There is a bug here where we scan the expected test results files: https://github.com/nexB/scancode-toolkit/tree/152abdaa73d1b8203a6f3cc6057d6c31c7c49e2b/tests/packagedcode/data/about
The directory could be reorged with a subdir with only .about files, this way the expected files are not part of the scan

pombredanne commented on tests/packagedcode/data/plugin/mui-package-expected.json:
```diff
@@ -18,11 +19,12 @@
       }
     }
   ],
+  "packages": [],
```
There should be something here: a simple way to check would be to verify that there are as many `package_manifests` as there are `packages`

pombredanne commented on src/scancode/api.py:
```diff
@@ -287,19 +287,22 @@ def _licenses_data_from_match(
 
 
-def get_package_info(location, **kwargs):
+def get_package_manifest_info(location, **kwargs):
```
Let's not rename... instead keep the old one and add a new one named "get_package_manifest"
And add a deprecation warning on the old one. See https://github.com/nexB/debian-inspector/blob/d0d3a5b0d08fd08493e73f0624d9c2758012b156/src/debut/copyright.py#L12

pombredanne commented on src/summarycode/score.py:
```diff
@@ -166,31 +166,31 @@ def compute_license_score(codebase):
 
 def get_declared_license_keys(codebase):
     """
-    Return a list of declared license keys found in packages and key files.
```
Let's avoid changing this

pombredanne commented on src/summarycode/summarizer.py:
```diff
@@ -112,7 +112,7 @@ def summarize_codebase(codebase, keep_details, **kwargs):
         ('holders', holder_summarizer),
```
Let's avoid most changes in summarycode if possible: conceptually the summaries are for packages, not for manifests

pombredanne commented on src/summarycode/summarizer.py:
```diff
@@ -112,7 +112,7 @@ def summarize_codebase(codebase, keep_details, **kwargs):
         ('holders', holder_summarizer),
         ('authors', author_summarizer),
         ('programming_language', language_summarizer),
-        ('packages', package_summarizer),
```
For instance here, do not rename "packages" to "package_manifests", instead populate "packages" from the new top-level "packages" attribute.

pombredanne commented on src/summarycode/summarizer.py:
```diff
@@ -408,53 +408,53 @@ def summarize_codebase_by_facet(codebase, **kwargs):
-    creatig the the "files" attribute from the `resource`. Yield back the
-    packages.
+    Update in-place every package mapping in the `package_manifests` list
+    by updating or creatig the the "files" attribute from the `resource`.
```
nit: creating

pombredanne reviewed (changes requested):
Thanks: see my review in line!

pombredanne reviewed (changes requested):
Also please update the changelog and bump the data version to 2.0 
You can hold on for the doc changes, and may have separate issue for the doc to do it at once.

AyanSinhaMahapatra committed:
Restore summarycode and API changes

Restores previously made changes in summarycode, and deprecates
package API with a warning, and adds other small bugfixes.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix packages bug for single file resources

Package_manifests were not being collected when there was
a single file in the codebase, this fixes the bug and regenerates tests.

Signed-off-by: Ayan Sinha <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Bump output data version and update CHANGELOG

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@pombredanne changes requested in the review are addressed now.

pombredanne commented on CHANGELOG.rst:
```diff
@@ -50,18 +53,20 @@ License detection:
 Package detection:
 ~~~~~~~~~~~~~~~~~~
 
-- We now support new package manifest formats:
```
Please keep existing planned updates in the changelog

pombredanne commented on src/scancode/api.py:
```diff
@@ -288,29 +288,68 @@ def _licenses_data_from_match(
+            # attention: we are swallowing ALL exceptions here!
+            pass
+
+    return dict(package_manifests=[])
```
Please keep the same API as before... the only addition to this function should be using the new recognize_package_manifests function and the deprecation warning.
It would be good to also add some  statement in the doc string that warns that the function is deprecated.

pombredanne reviewed (changes requested):
Thank you for making the changes,
Some extra things for your review before we can merge.

pombredanne reviewed (changes requested):
(The previous comment was made on the wrong PR )

AyanSinhaMahapatra commented:
All green now.

pombredanne reviewed (approved):
LGTM... merging: thank you!
